app/src/routes/api/v1/dataset.router.js

```diff
@@ -197,83 +198,92 @@ class DatasetRouter {
+                    if (
+                        sort.indexOf('metadata') >= 0
+                        || sort.indexOf('relevance') >= 0
+                    ) {
+                        metadataSort = sort;
+                    }
+
+                    const metadataIds = await RelationshipsService.filterByMetadata(search, metadataSort);
+                    const searchBySynonymsIds = await RelationshipsService.searchBySynonyms(serializeObjToQuery(query));
+                    const datasetBySearchIds = await DatasetService.getDatasetIdsBySearch(search.split(' '));
+                    searchIds = metadataIds.concat(searchBySynonymsIds).concat(datasetBySearchIds);
+                }
+                if (
+                    serializeObjToQuery(query).indexOf('concepts[0][0]') >= 0
+                    || sort.indexOf('most-favorited') >= 0
+                    || sort.indexOf('most-viewed') >= 0
+                ) {
+                    conceptIds = await RelationshipsService.filterByConcepts(serializeObjToQuery(query));
+                }
+                if ((searchIds && searchIds.length === 0) || (conceptIds && conceptIds.length === 0)) {
+                    ctx.body = DatasetSerializer.serialize([], null);
+                    return;
+                }
+                const finalIds = searchIds && conceptIds ? arrayIntersection(conceptIds, searchIds) : searchIds || conceptIds;
+                const uniqueIds = new Set([...finalIds]); // Intersect and unique
+                ctx.query.ids = [...uniqueIds].join(); // it has to be string
             }
-            if ((searchIds && searchIds.length === 0) || (conceptIds && conceptIds.length === 0)) {
-                ctx.body = DatasetSerializer.serialize([], null);
-                return;
+            // Links creation
+            const clonedQuery = Object.assign({}, query);
+            delete clonedQuery['page[size]'];
+            delete clonedQuery['page[number]'];
+            delete clonedQuery.ids;
+            const serializedQuery = serializeObjToQuery(clonedQuery) ? `?${serializeObjToQuery(clonedQuery)}&` : '?';
+            const apiVersion = ctx.mountPath.split('/')[ctx.mountPath.split('/').length - 1];
+            const link = `${ctx.request.protocol}://${ctx.request.host}/${apiVersion}${ctx.request.path}${serializedQuery}`;
+            const datasets = await DatasetService.getAll(query, user && user.role === 'ADMIN');
+            ctx.set('cache', `dataset ${query.includes ? query.includes.split(',').map(elem => elem.trim()).join(' ') : ''}`);
+            ctx.body = DatasetSerializer.serialize(datasets, link);
+        } catch (err) {
+            if (err instanceof InvalidRequest) {
+                ctx.throw(400, err);
             }
-            // searchIds = searchIds || [];
-            // conceptIds = conceptIds || [];
-            const finalIds = searchIds && conceptIds ? arrayIntersection(searchIds, conceptIds) : searchIds || conceptIds;
```

Comment:
This is one of the bugs I found: our implementation of `arrayIntersection` filters arg1 by values of arg2, so we retain the order of the first argument - which should then be conceptIds, not searchIds